sudoku_solver/sudoku/test/Spec.hs

```diff
@@ -0,0 +1,59 @@
+module Main where
+
+import Test.HUnit
+import Data.List (nub)
+import qualified Data.Vector as V
+
+import Board
+import Solver (solve)
+import SampleBoards (board1, board1Solved, invalidBoard, boardHard)
+
+isValidSolution :: Board -> Bool
+isValidSolution board = all isValidRow board && all isValidColumn [0..8] && all isValidSubgrid [0..2]
+  where
+    isValidRow row = let values = V.toList row in length (filter (/= 0) values) == length (nub values)
+    isValidColumn colIndex = let values = map (V.! colIndex) (V.toList board) in length (filter (/= 0) values) == length (nub values)
+    isValidSubgrid gridIndex = let subgrid = [ board V.! (row + 3 * (gridIndex `div` 3)) V.! (col + 3 * (gridIndex `mod` 3)) | row <- [0..2], col <- [0..2]]
+      in length (filter (/= 0) subgrid) == length (nub subgrid)
```

Comment:
Why? You already have `isValid` function in the implementation

Reply:
Because those are different, this one is to check that the current state is valid, and the one implemented in Solver.hs is to check that the new move does result in a valid board

Reply:
Oh okay I guess second one can be implemented using the first one, makes sense